enclave_build/src/image_manager.rs

```diff
@@ -0,0 +1,144 @@
+use crate::storage::OciStorage;
+use crate::{EnclaveBuildError, Result};
+
+use tokio::runtime::Runtime;
+
+pub struct OciImageManager {
+    /// Name of the container image.
+    image_name: String,
+    /// Have the storage as an option for the CLI to continue running if there is a storage creation
+    /// error (the image will simply be pulled in this case, instead of being fetched from storage)
+    storage: Option<OciStorage>,
+}
+
+impl OciImageManager {
+    /// When calling this constructor, it also tries to initialize the storage at the default path.
+    /// If this fails, the ImageManager is still created, but the 'storage' field is set to 'None'.
+    pub fn new(image_name: &str) -> Self {
+        // Add the default ":latest" tag if the image tag is missing
+        let image_name = check_tag(image_name);
+
+        // The docker daemon is not used, so a local storage needs to be created
+        // Get the default storage root path
+        let root_path = match OciStorage::get_default_root_path() {
+            Ok(path) => path,
+            Err(_) => {
+                // If the storage root path could not be determined, then the storage can not be initialized
+                return Self {
+                    image_name,
+                    storage: None,
+                };
+            }
+        };
+
+        // Try to create/read the storage
+        let storage = match OciStorage::new(&root_path) {
+            Ok(manager) => Some(manager),
+            Err(err) => {
+                // If the storage could not be created, log the error
+                eprintln!("{:?}", err);
+                None
+            }
+        };
+
+        Self {
+            image_name,
+            storage,
+        }
+    }
+
+    /// Returns a struct containing image metadata.
+    ///
+    /// If the image is stored correctly, the function tries to fetch the image from the storage.
+    ///
+    /// If the image is not stored or a storage was not created (the 'storage' field is None),
+    /// it pulls the image, stores it (if the 'storage' field is not None) and returns its metadata.
+    ///
+    /// If the pull succeeded but the store operation failed, it returns the pulled image metadata.
+    async fn get_image_details(&mut self, image_name: &str) -> Result<ImageDetails> {
+        let image_name = check_tag(image_name);
+
+        let local_storage = (self.storage).as_mut();
+
+        if let Some(storage) = local_storage {
+            // Try to fetch the image from the storage
+            let image_details = storage.fetch_image_details(&image_name).map_err(|err| {
+                // Log the fetching error
+                eprintln!("{:?}", err);
+                err
+            });
+
+            // If the fetching failed, pull it from remote and store it
+            return image_details.or(self.fetch_and_try_store_image(&image_name).await);
+        }
+
+        self.fetch_and_try_store_image(&image_name).await
```

Comment:
Can we refactor this a bit so that we call `fetch_and_try_store_image` only once?